aten/src/ATen/core/ivalue.cpp

```diff
@@ -940,17 +940,24 @@ TORCH_API intrusive_ptr<ivalue::Future> collectAll(
   };
 
   auto ctx = std::make_shared<Ctx>(std::move(srcs));
-  std::function<void()> func = [ctx]() {
-    if (--ctx->remaining == 0) {
-      ctx->dstFuture->markCompleted(ctx->asIvalue);
-    }
-  };
   if (ctx->srcFutures.size() == 0) {
     ctx->dstFuture->markCompleted(ctx->asIvalue);
   } else {
     auto typePtr = ctx->srcFutures.get(0)->elementType();
     for (const auto i : c10::irange(ctx->srcFutures.size())) {
-      TORCH_CHECK(i == 0 || *ctx->srcFutures.get(i)->elementType() == *typePtr);
```

Comment:
I actually removed this on purpose since I felt the check doesn't make sense? The check is asserting that all the Futures contain the same type which seems to be too restrictive since you should be able to wait on a collection of futures irrespective of the underlying type they return. This was actually causing one of my tests to fail where one Future returned a PyObject and the other a Tensor.

Reply:
Are you sure this will not create problems with TorchScript's type system? I though it wasn't able to handle non-homogeneous types, by which I mean that it requires a list to have type `List[T]`, hence have each item of the same type `T`. I don't know if in case of futures this requirement is relaxed, as it seems that future are a somewhat type-erased class... (in other words, in TorchScript's view, is this a `List[Future]` or a `List[Future[T]]`?)

If in your case you're expecting a mix of Tensors and PyObjects couldn't you solve it by using `Future[Union[Tensor, PyObject]]`?

Reply:
My understanding is that TorchScript sees this as `List[Future]` and not `List[Future[T]]`.

> If in your case you're expecting a mix of Tensors and PyObjects couldn't you solve it by using Future[Union[Tensor, PyObject]]?

I'm not sure where exactly I would make this change? My unit test is completely in python and generates a PyObject for one rpc_async and a Tensor for the other.

Reply:
> I'm not sure where exactly I would make this change? My unit test is completely in python and generates a PyObject for one rpc_async and a Tensor for the other.

Uhm, yeah, good point, the type of a future is determined by its creator, and we don't seem to have a way to change it or "cast it" to something else...